### ISSUE TITLE
Different Enumerator Behavior from MRI 1.9.3, 2.0.0 and 2.1.0

### ISSUE THREAD

**Issue by dlecocq** [jruby/jruby#1430]
The following snippet behaves differently in JRuby 1.7.9 and 1.7.10 from MRI 1.9.3 and 2.0.0. It features an `Enumerator` that pops from a set of queues in round-robin fashion and a helper function that accumulates popped items until `count` is reached or the queues are exhausted.

To test it, I have a pop to ensure that it's empty to begin with. Then I push an item and then try to pop it. In JRuby, I have to pop a couple of times after adding the item to get the result. In MRI, it's available immediately after pushing it into the queues.

``` ruby
#! /usr/bin/env ruby

@queues = 10.times.map { Array.new }

# An enumerator that yields `nil` if none of the queues has any items, but
# otherwise round-robins across them.
@foo = Enumerator.new do |enum|
  loop do
    found = false
    @queues.each do |queue|
      job = queue.pop
      unless job.nil?
        found = true
        enum.yield(job)
      end
    end
    enum.yield(nil) unless found
  end
end

# Pop up to 'count' objects from 
def pop(count)
  jobs = []
  while jobs.length < count
    job = @foo.next
    break if job.nil?
    jobs.push(job)
  end
  jobs
end

# First, verify that the there are no items available
puts "Pop before put: #{pop(5)}"
@queues[5].push(1)
# Now, I'd expect the item to be immediately available after putting it, but in
# jRuby, it's not yielded until the second pop.
puts "First pop after put: #{pop(5)}"
puts "Second pop after put: #{pop(5)}"

# jruby 1.7.9 (1.9.3p392) 2013-12-06 87b108a on Java HotSpot(TM) 64-Bit Server VM 1.7.0_45-b18 [darwin-x86_64]
# Pop before put: []
# First pop after put: []
# Second pop after put: [1]

# ruby 1.9.3p429 (2013-05-15 revision 40747) [x86_64-darwin13.0.0]
# Pop before put: []
# First pop after put: [1]
# Second pop after put: []

# ruby 2.0.0p247 (2013-06-27 revision 41674) [universal.x86_64-darwin13]
# Pop before put: []
# First pop after put: [1]
# Second pop after put: []

# ruby 2.1.0dev (2013-12-20 trunk 44301) [x86_64-darwin13.0]
# Pop before put: []
# First pop after put: [1]
# Second pop after put: []
```

**Comment by kares** [comment]
jruby 9.1.9.0 (2.3.3)   64-Bit Server VM 25.121-b13 on 1.8.0_121-b13 +jit [linux-x86_64]
```
Pop before put: []
First pop after put: []
Second pop after put: [1]
```

**Comment by kares** [comment]
^^ seems acceptable to me (JRuby 1.7.x will stay as is)

**Comment by headius** [comment]
This matches MRI in current JRuby. Calling it fixed as of 9.2.7.

```
$ jruby blah.rb
Pop before put: []
First pop after put: [1]
Second pop after put: []
```
